File: 213-house-robber-ii/house-robber-ii.java
class Solution {
    int[] dp;
    public int rob(int[] nums) {
        dp=new int[nums.length];
        int n=nums.length;
        if(n==1){
            return nums[0];
        }
        Arrays.fill(dp,-1);
        int ans=robber(nums,0,n-2);
        Arrays.fill(dp,-1);
        int ans2=robber(nums,1,n-1);
        return Math.max(ans,ans2);
    }
    public int robber(int[] nums,int si,int ei){
        if(si>ei){
            return 0;
        }
        if(dp[si]!=-1){
            return dp[si];
        }
        int rob=nums[si]+robber(nums,si+2,ei);
        int skip=robber(nums,si+1,ei);
        return dp[si]=Math.max(rob,skip);
    }
}

File: 213-house-robber-ii/house-robber-ii.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n=nums.size();
        if(n==1){
            return nums[0];
        }
        vector<int> dp(n,-1);
        int ans1=robber(nums,0,n-2,dp);
        dp.assign(n,-1);
        return max(ans1,robber(nums,1,n-1,dp));
    }
    int robber(vector<int> &nums,int idx,int n,vector<int> &dp){
        if(idx>n){
            return 0;
        }
        if(dp[idx]!=-1){
            return dp[idx];
        }
        int rob=nums[idx]+robber(nums,idx+2,n,dp);
        int nrob=robber(nums,idx+1,n,dp);
        return dp[idx]=max(rob,nrob);
    }
};